File: sol/solution/0001-0100/0024. Swap Nodes in Pairs/Solution.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
    ListNode? swapPairs(ListNode? head) {
        if (head == null || head.next == null) return head;
        final dummy = ListNode(0, head);
        ListNode? prev = dummy;
        while (prev!.next != null && prev.next!.next != null) {
            final first = prev.next!;
            final second = first.next!;
            first.next = second.next;
            second.next = first;
            prev.next = second;
            prev = first;
        }
        return dummy.next;
    }
}

File: sol/solution/0001-0100/0024. Swap Nodes in Pairs/Solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode { val: 0, next: head });
        let mut prev = &mut dummy;
        while let Some(mut first) = prev.next.take() {
            if let Some(mut second) = first.next.take() {
                first.next = second.next.take();
                second.next = Some(first);
                prev.next = Some(second);
                prev = prev.next.as_mut().unwrap().next.as_mut().unwrap();
            } else {
                prev.next = Some(first);
                break;
            }
        }
        dummy.next
    }
}